spicy/src/compiler/visitors/validator.cc

```diff
@@ -224,14 +224,21 @@ struct PreTransformVisitor : public hilti::visitor::PreOrder<void, PreTransformV
         if ( f.parseType().isA<type::Bytes>() && ! f.ctor() ) {
             auto eod_attr = AttributeSet::find(f.attributes(), "&eod");
             auto until_attr = AttributeSet::find(f.attributes(), "&until");
+            auto until_including_attr = AttributeSet::find(f.attributes(), "&until-including");
 
-            if ( eod_attr ) {
-                if ( until_attr )
-                    error("&eod incompatible with &until", p);
+            int num_attrs = 0;
+            for ( const auto& i : {eod_attr, parse_from_attr, parse_at_attr, until_attr, until_including_attr} ) {
+                if ( i )
+                    ++num_attrs;
             }
 
-            else if ( ! until_attr && ! size_attr && ! parse_from_attr && ! parse_at_attr )
-                error("bytes field requires one of &size, &eod, or &until", p);
+            // &size can be combined with any other attribute, or be used standalone.
+
+            if ( num_attrs > 1 )
+                error("incompatible attributes for bytes field", p);
```

Comment:
I'd imagine it might be hard for users to figure out their error here.

How instead of `num_attrs` we store the present attributes (e.g., in a `vector<Attribute>` or `vector<Attribute*>` if you want to prevent allocations), so we can show them to the user, e.g.,
```
if (attrs.size() > 1)
  error(fmt("attributes cannot be combined: %s", join(attrs, ","));
```

Reply:
I like that, will do.